[#3] ContextAPI에서 Jotai로 변경 후 다크모드 적용

kjheee opened the pull request:
## PR 타입
- [X] 패키지 추가
- [X] 코드 개선

## 작업 내용
1. Context 폴더 삭제 및 관련 코드 제거
2. Jotai atom으로 다크모드 구현

## 관련 이슈
close #3

kjheee committed:
#3 chore: Jotai 패키지 설치

kjheee committed:
#3 remove: Context 폴더 삭제

kjheee committed:
#3 refactor: Context Provider 코드 제거

kjheee committed:
#3 refactor: useTheme 커스텀훅에 useAtom 적용

- useState 대체

kjheee committed:
#3 refactor: Header 컴포넌트에서 useTheme 사용

- useDarkMode 대체

f-lab-jake commented on src/hooks/useTheme.ts:
```diff
@@ -1,15 +1,18 @@
-    const [darkMode, setDarkMode] = useState(initialState);
+const darkModeAtom = atom(false);
+
+export const useTheme = (): DarkModeState => {
```
훅을 굳이 안만드려고 jotai를 씁니다.

f-lab-jake replied:
https://velog.io/@taez224/Jotai-Tutorial-Write-Only-atom

kjheee committed:
#3 remove: useTheme 훅 제거

kjheee committed:
#3 chore: tsconfig.json 수정

- store 절대경로 설정

kjheee committed:
#3 refactor: darkMode atom 생성 및 적용

- 상태를 변경할 때 updateDarkMode 함께 호출
- [App] 로컬 테마 정보 조회
- [Header] useTheme -> useAtom 변경

kjheee commented:
기존 커스텀훅을 제거하고 atom으로 다시 설정했습니다.
1. 상태를 set할 때, 로컬 테마 정보와 css 클래스를 변경하는 함수를 함께 호출합니다.
2. 렌더링 시 테마 상태를 유지하는 코드는 App 컴포넌트에서 구현합니다.

f-lab-jake commented on src/App.tsx:
```diff
@@ -1,7 +1,18 @@
 
 function App() {
+    const setDarkMode = useSetAtom(darkMode_atom);
+    useEffect(() => {
+        const isDark =
+            localStorage.theme === 'dark' ||
+            (!('theme' in localStorage) &&
+                window.matchMedia('(prefers-color-scheme: dark)').matches);
+        setDarkMode(isDark);
+    }, [setDarkMode]);
```
여전히 분리하지 못했습니다.
[`atomWithStorage`](https://jotai.org/docs/utilities/storage)를 참고하세요.

f-lab-jake commented on src/store/atoms/theme.ts:
```diff
@@ -0,0 +1,20 @@
+);
+
+const updateDarkMode = (darkMode: boolean) => {
+    if (darkMode) {
+        document.documentElement.classList.add('dark');
+        localStorage.theme = 'dark';
+    } else {
+        document.documentElement.classList.remove('dark');
+        localStorage.theme = 'light';
+    }
```
렌더링 로직이죠 이건?

kjheee committed:
#3 refactor: base atom을 atomWithStorage로 변경

- App 컴포넌트: localStorage 테마 정보 조회 및 스타일 변경 코드 삭제

kjheee committed:
#3 feat: 테마 토글 버튼 컴포넌트 분리

kjheee committed:
#3 refactor: 분리한 테마 버튼 컴포넌트 적용

kjheee commented:
1. Base atom을 `atomWithStorage`로 변경했습니다.
2. 테마 토글 버튼을 컴포넌트로 만들고 내부에서 useEffect를 사용했습니다.
-class를 변경하는 함수는 컴포넌트 밖으로 빼내었습니다.

f-lab-jake commented on src/components/UI/ThemeButton/ThemeButton.tsx:
```diff
@@ -0,0 +1,36 @@
+const cx = classnames.bind(styles);
+
+const toggleDarkClass = (darkMode: boolean) => {
+    if (darkMode) {
+        document.documentElement.classList.add('dark');
+    } else {
+        document.documentElement.classList.remove('dark');
+    }
```
렌더링 로직을 react 바깥에 넣는걸 피하세요

kjheee replied:
넵! 변경했습니다.

kjheee committed:
#3 refactor: 렌더링 로직 컴포넌트 내부 선언으로 변경

f-lab-jake commented on src/components/UI/ThemeButton/ThemeButton.tsx:
```diff
@@ -6,21 +6,21 @@ import { IoIosSunny, IoMdMoon } from 'react-icons/io';
         toggleDarkClass(!darkMode);
     };
 
+    const toggleDarkClass = (darkMode: boolean) => {
+        if (darkMode) {
+            document.documentElement.classList.add('dark');
+        } else {
+            document.documentElement.classList.remove('dark');
+        }
+    };
```
`document` 직접 dom api로 조작하지 말란 이야기입니다.

f-lab-jake replied:
지금은 위치만 바뀌었어요.

kjheee committed:
#3 refactor: 테마 적용 방법 dataset으로 변경

kjheee committed:
chore: jotai 재설치

kjheee committed:
remove: Root 관련 파일 삭제

kjheee committed:
#3 refactor: Root 페이지를 App 컴포넌트로 변경 및 다크모드 dataset 적용

- router root 컴포넌트 변경
- Root 관련 스타일 코드 index.css에 작성

kjheee committed:
#3 refactor: 렌더링 로직 삭제

kjheee commented:
멘토님, 기존 Root 페이지를 제거하고 관련 코드를 App 컴포넌트에 작성했습니다. 
- App 컴포넌트
   - 최상위 태그에 다크모드 data-theme 적용
   - 기존 RootContainer 스타일 적용(index.css)

처음에는 body에 dataset을 적용하도록 수정했었습니다. [0ca330f](https://github.com/f-lab-edu/debus/pull/4/commits/0ca330f5bbcb9586edab9b337d5e9838e348b973) 이후 커밋부터 확인 부탁드립니다!

f-lab-jake commented on src/components/UI/ThemeButton/ThemeButton.tsx:
```diff
@@ -0,0 +1,23 @@
+
+const ThemeButton = () => {
+    const [darkMode, setDarkMode] = useAtom(darkMode_atom);
+    const handleTheme = () => {
+        setDarkMode(!darkMode);
+    };
```
이런 함수는 의미가 없습니다. 이름으로 뭘 알려주는 것도 아니죠. 그냥 props에 인라인으로 쓰는게 낫습니다.

kjheee replied:
넵, 인라인으로 작성했습니다.

f-lab-jake commented on src/store/atoms/theme.ts:
```diff
@@ -0,0 +1,10 @@
+import { atomWithStorage } from 'jotai/utils';
+
+export const darkMode_storage_atom = atomWithStorage('darkMode', false);
+export const darkMode_atom = atom(
+    (get) => get(darkMode_storage_atom),
+    (_, set, darkMode: boolean) => {
+        set(darkMode_storage_atom, darkMode);
+    },
+);
```
이거 쓸필요 없이 그냥 `darkModeAtom`을 그대로 쓰면 됩니다. 보시면 얘는 하는게 없어요.

f-lab-jake replied:
변수명은 일관성있게 camel case를 써주세요. 다만 atom같은 좀 특이한 그룹에 대해서 다른 룰을 적용하는 것도 의미가 있긴합니다.

f-lab-jake replied:
`darkMode_atom`을 굳이 추가할거면, write에 toggle 동작을 넣는걸 고려해볼만합니다.

kjheee replied:
- 처음엔 atom이 여러 개가 만들어져서 구분이 잘 되지 않아 다른 네이밍 규칙을 적용했는데 개선하면서 큰 의미가 없어졌습니다. 일관성 있게 작성하겠습니다.
- write에 toggle을 적용하니 사용 시 set함수에 인자를 넣지 않아도 되어서 말씀해주신 방법을 적용해보았습니다.

f-lab-jake commented on src/App.tsx:
```diff
@@ -1,8 +1,18 @@
-    return <RouterProvider router={router} />;
+    const darkMode = useAtomValue(darkMode_atom);
+    return (
+        <div id="app" data-theme={darkMode ? 'dark' : 'light'}>
```
넵 이런식으로 하는게 맞습니다. 그리고 그냥 class로 주는게 어떤가요?  data- 로 해놓으면 selector 쓸때 좀 귀찮을수도 있어요. 혹시 참고한 예시가 있다면  공유해주세요.

kjheee replied:
이전에는 html 또는 body에 직접 테마 상태를 지정하는 참고 자료를 보고 dataset을 고려했어요.
-참고 1: https://lukelowrey.com/css-variable-theme-switcher/
-참고 2: https://jeonghwan-kim.github.io/dev/2021/05/17/css-variable.html

- App 컴포넌트는 module.css를 만들지 않고, index.css에서 id 선택자를 이용하고 있어서 목적이 구분되면 좋을 것 같아 dataset을 적용했었습니다.

kjheee committed:
#3 refactor: 다크모드 스타일 적용 방법 dataset -> class로 변경

kjheee committed:
#3 refactor: atom 변수명 camelCase로 변경 & write atom toggle 적용

f-lab-jake commented on src/components/UI/ThemeButton/ThemeButton.tsx:
```diff
@@ -1,19 +1,16 @@
-    const handleTheme = () => {
-        setDarkMode(!darkMode);
-    };
+    const [darkMode, setDarkMode] = useAtom(darkModeAtom);
```
이렇게할때는 이름을 `toggleDarkMode`로 하는게 낫겠죠

kjheee replied:
넵! 네이밍 다시 한 번 체크하겠습니다. 변경 후 머지하겠습니다!
